### PR TITLE
fix(zone.js): store remove abort listener on the scheduled task

### LINKED ISSUE #56148: AbortController with zone.js not working anymore
### Which @angular/* package(s) are the source of the bug?

zone.js

### Is this a regression?

Yes

### Description

Since version 0.14.6 of zone.js, the functionality of the AbortController seems broken.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/stackblitz-starters-dq9ojm?file=src%2Fmain.ts

### Please provide the exception or error you saw

```true
With zone.js 0.14.5 the AbortController was working correctly, but with 0.14.6 it started to be broken.

Reproduction:
1. Open https://stackblitz.com/edit/stackblitz-starters-dq9ojm?file=src%2Fmain.ts
2. Click on "Click" button
3. Alert message is presented
4. Click on "Abort" button
5. Click on "Click" button and the alert message is still presented, although the abort() method of the AbortController was called.
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 18.0.1
Node: 18.20.3
Package Manager: npm 10.2.3
OS: linux x64

Angular: 18.0.0
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1800.1
@angular-devkit/build-angular   18.0.1
@angular-devkit/core            18.0.1
@angular-devkit/schematics      18.0.1
@angular/cli                    18.0.1
@schematics/angular             18.0.1
rxjs                            7.8.1
typescript                      5.4.5
zone.js                         0.14.6
```

### Anything else?

Related PR which probably broke the behavior: #55339

### PR DESCRIPTION
**Pull request by arturovt**
Prior to this commit, a memory leak occurred when the `abort` listener was not removed from the `AbortSignal`. We introduced a fix to remove the event listener, but it was erroneously stored on the `taskData`, which is a shared global object. Consequently, when something attempted to remove an event listener, it immediately removed the last stored abort listener. As a result, events would never be canceled.

We have now rectified this by storing the remove abort listener function directly on the task itself. This adjustment ensures that the abort listener is tied only to the specific task. When the `abort` function is called, it cancels the task. Therefore, it is safe to associate the cleanup function directly with the task.

Closes: #56148

### CODE REVIEW

File: packages/zone.js/lib/common/events.ts [hunk 1]
```diff
@@ -693,15 +729,10 @@ export function patchEventTarget(
...
-
-            // Note that `removeAllListeners` would ultimately call `removeEventListener`,
-            // so we're safe to remove the abort listener only once here.
-            const taskData = existingTask.data as EventTaskData;
-            if (taskData?.removeAbortListener) {
-              taskData.removeAbortListener();
-              taskData.removeAbortListener = null;
-            }
-
+            // Under normal conditions, this will trigger `cancelFn` on the `existingTask`,
```
Comment (AndrewKushnir) [inline packages/zone.js/lib/common/events.ts:732]: Just curious if there are some conditions when this is not the case? Should we document that here as well?

**Review by AndrewKushnir — COMMENTED** [review]
@arturovt thanks for the fix, I've just left a minor comment.

@JiaLiPassion could you please take a look at this fix when you get a chance?

File: packages/zone.js/lib/common/events.ts (hunk 1 above)
Comment (arturovt) [reply packages/zone.js/lib/common/events.ts]: I've updated the comments and attached the old issue that existed a long time ago in 2017.

**Review by AndrewKushnir — APPROVED** [review]
@arturovt thanks for addressing the feedback 👍

**Review by JiaLiPassion — APPROVED** [review]
LGTM! Thanks.

**Comment by AndrewKushnir** [comment]
[Global Presubmit](https://test.corp.google.com/OCL:639981153:BASE:639983257:1717466843646:265952c9).

**Comment by AndrewKushnir** [comment]
**Caretaker note**: TGP is "green", this PR is ready for merge.

**Comment by thePunderWoman** [comment]
This PR was merged into the repository by commit 4a3800a6a0ae9d667dd961c6e4029c01c6819988.

**Pull request closed without merging**